src/connector/src/source/filesystem/opendal_source/opendal_reader.rs

```diff
@@ -182,3 +224,101 @@ impl<Src: OpendalSource> OpendalReader<Src> {
+
+// The hidden columns that must be included here are _rw_file and _rw_offset.
+// Depending on whether the user specifies a primary key (pk), there may be an additional hidden column row_id.
+// Therefore, the maximum number of hidden columns is three.
+const MAX_HIDDEN_COLUMN_NUMS: usize = 3;
+fn convert_record_batch_to_stream_chunk(
+    record_batch: RecordBatch,
+    source_columns: Option<Vec<Column>>,
+    file_name: String,
+) -> Result<StreamChunk, crate::error::ConnectorError> {
+    match source_columns {
+        Some(source_columns) => {
+            let size = source_columns.len();
+            let mut chunk_columns =
+                Vec::with_capacity(source_columns.len() + MAX_HIDDEN_COLUMN_NUMS);
+            for source_column in source_columns {
+                if let Some(parquet_column) = record_batch.column_by_name(&source_column.name) {
+                    let converted_arrow_data_type =
+                        arrow_schema::DataType::try_from(&source_column.data_type)?;
+
+                    if &converted_arrow_data_type == parquet_column.data_type() {
+                        let column = Arc::new(parquet_column.try_into()?);
+                        chunk_columns.push(column);
+                    } else {
+                        // data type mismatch, this column is set to null.
+                        let mut array_builder =
+                            ArrayBuilderImpl::with_type(size, source_column.data_type);
+
+                        array_builder.append_n_null(record_batch.num_rows());
+                        let res = array_builder.finish();
+                        let column = Arc::new(res);
+                        chunk_columns.push(column);
+                    }
+                } else if !source_column.is_visible {
+                    // For hidden columns in the file source:
+                    // - The `row_id` column, which will be filled by the row ID generator, is left as null.
+                    // - The `_rw_file` column is filled with the object name.
+                    // - The `_rw_offset` column is filled with the offset.
+                    let mut array_builder =
+                        ArrayBuilderImpl::with_type(size, source_column.data_type);
+                    // Hidden column naming rules refer to https://github.com/risingwavelabs/risingwave/blob/3736982d6fe648ad32b50e1d3ce97119edcab1a6/src/connector/src/parser/additional_columns.rs#L289
+                    let datum: Datum = if source_column.name.ends_with("file") {
+                        Some(ScalarImpl::Utf8(file_name.clone().into()))
+                    } else if source_column.name.ends_with("offset") {
+                        // `_rw_offset` is not used.
+                        Some(ScalarImpl::Utf8("0".into()))
+                    } else {
+                        // row_id column.
+                        None
+                    };
```

Comment:
Though the current implementation based on column name matching is correct, I found it a littble bit hacky, given that we have a strongly typed `addtional_column.column_type`. Is it possible to adopt what we have done in [`SourceStreamChunkRowWriter`](https://github.com/risingwavelabs/risingwave/blob/dde79524552a3dc02f6055bcd1bdb531f1e763c0/src/connector/src/parser/mod.rs#L385) for the additional columns? Maybe we can extract the common logic to avoid duplicate codes.

Furthermore, maybe we can add a `col_writer` method in [`SourceStreamChunkBuilder`](https://github.com/risingwavelabs/risingwave/blob/dde79524552a3dc02f6055bcd1bdb531f1e763c0/src/connector/src/parser/mod.rs#L108) to convert columnar record batch to columnar stream chunk as well. Just an idea. Not sure whether it can make things easier.